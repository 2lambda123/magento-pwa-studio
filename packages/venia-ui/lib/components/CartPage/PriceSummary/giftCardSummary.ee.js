import React, { Fragment } from 'react';
import { Price } from '@magento/peregrine';

import { mergeClasses } from '../../../classify';

const DEFAULT_AMOUNT = {
    currency: 'USD',
    value: 0
};

/**
 * Reduces applied gift card amounts into a single amount.
 *
 * @param {Array} cards
 */
const getGiftCards = (cards = []) => {
    if (!cards.length) {
        return DEFAULT_AMOUNT;
    } else {
        return {
            currency: 'USD',
            value: cards.reduce(
                (acc, card) => acc + card.applied_balance.value,
                0
            )
        };
    }
};

/**
 * A component that renders the gift card summary line item.
 *
 * @param {Object} props.classes
 * @param {Object} props.data fragment response data
 */

export default props => {
    const classes = mergeClasses({}, props.classes);
    const cards = getGiftCards(props.data);

    return cards.value ? (
        <Fragment>
            <span className={classes.lineItemLabel}>
                {'Gift Card(s) applied'}
            </span>
            <span className={classes.price}>
                {'-'}
                <Price value={cards.value} currencyCode={cards.currency} />
            </span>
        </Fragment>
    ) : null;
};
