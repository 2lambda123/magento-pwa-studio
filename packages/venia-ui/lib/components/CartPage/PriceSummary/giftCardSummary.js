import React from 'react';
import gql from 'graphql-tag';
import { Price } from '@magento/peregrine';

import { mergeClasses } from '../../../classify';

/**
 * Gift cards are an EE feature and we need some way to conditionally include
 * this code and query in bundled assets. For now, this constant will be the
 * toggle, but we may eventually need to do something like a build tool that can
 * check schema for existence of "applied_gift_cards" and include this component
 * if present.
 *
 * Right now, if `IS_EE == false`, this component returns a no-op, null function
 * as well as an "empty" fragment so as to not break the GQL query of a parent.
 *
 * TODO: Solve this problem. This local toggle should not be long-lived, at least not past PWA-78.
 */
const IS_EE = false;

const DEFAULT_AMOUNT = {
    currency: 'USD',
    value: 0
};

/**
 * Reduces applied gift card amounts into a single amount.
 *
 * @param {Array} cards
 */
const getGiftCards = (cards = []) => {
    if (!cards.length) {
        return DEFAULT_AMOUNT;
    } else {
        return {
            currency: 'USD',
            value: cards.reduce(
                (acc, card) => acc + card.applied_balance.value,
                0
            )
        };
    }
};

/**
 * A component that renders the gift card summary line item.
 *
 * @param {Object} props.classes
 * @param {Object} props.data fragment response data
 */
const GiftCardSummary = IS_EE
    ? props => {
          const classes = mergeClasses({}, props.classes);
          const cards = getGiftCards(props.data);

          return cards.value ? (
              <>
                  <span className={classes.lineItemLabel}>
                      {'Gift Card(s) applied'}
                  </span>
                  <span className={classes.price}>
                      {'-'}
                      <Price
                          value={cards.value}
                          currencyCode={cards.currency}
                      />
                  </span>
              </>
          ) : null;
      }
    : () => null;

export const GiftCardSummaryFragment = IS_EE
    ? gql`
          fragment GiftCardSummaryFragment on Cart {
              id
              applied_gift_cards {
                  applied_balance {
                      value
                      currency
                  }
              }
          }
      `
    : gql`
          fragment GiftCardSummaryFragment on Cart {
              id
              __typename
          }
      `;

export default GiftCardSummary;
