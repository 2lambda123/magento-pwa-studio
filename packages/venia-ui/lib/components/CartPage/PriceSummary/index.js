export { default, PriceSummaryQuery } from './priceSummary';
