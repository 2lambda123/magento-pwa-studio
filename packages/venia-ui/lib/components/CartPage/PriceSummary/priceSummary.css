.root {
    padding: 0 1rem 1rem 1rem;
}

.lineItems {
    display: grid;
    grid-row-gap: 0.75rem;
    grid-template-columns: 50% 50%;
    padding-bottom: 1rem;
    line-height: 1.5rem;
}

.lineItemLabel {
    justify-self: start;
}

.price {
    justify-self: end;
}

.totalLabel {
    composes: lineItemLabel;
    font-weight: bold;
}

.totalPrice {
    composes: price;
    font-weight: bold;
}

.checkoutButton_container {
    text-align: center;
}
