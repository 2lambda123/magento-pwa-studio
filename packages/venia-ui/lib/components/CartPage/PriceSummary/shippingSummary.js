import React from 'react';
import gql from 'graphql-tag';
import { Price } from '@magento/peregrine';

import { mergeClasses } from '../../../classify';
/**
 * A component that renders the shipping summary line item after address and
 * method are selected
 *
 * @param {Object} props.classes
 * @param {Object} props.data fragment response data
 */
const ShippingSummary = props => {
    const classes = mergeClasses({}, props.classes);
    const { data, isCheckout } = props;

    // Don't render estimated shipping until an address has been provided and
    // a method has been selected.
    if (!data.length || !data[0].selected_shipping_method) {
        return null;
    }

    const shipping = data[0].selected_shipping_method.amount;

    // For a value of "0", display "FREE".
    const price = shipping.value ? (
        <Price value={shipping.value} currencyCode={shipping.currency} />
    ) : (
        <span>{'FREE'}</span>
    );

    return (
        <>
            <span className={classes.lineItemLabel}>
                {isCheckout ? 'Shipping' : 'Estimated Shipping'}
            </span>
            <span className={classes.price}>{price}</span>
        </>
    );
};

export const ShippingSummaryFragment = gql`
    fragment ShippingSummaryFragment on Cart {
        id
        shipping_addresses {
            selected_shipping_method {
                amount {
                    currency
                    value
                }
            }
        }
    }
`;

export default ShippingSummary;
