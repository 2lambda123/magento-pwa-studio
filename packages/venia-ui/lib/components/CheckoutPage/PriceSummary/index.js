import React from 'react';

export default () => <div>Price Summary Will be handled in PWA-184</div>;
