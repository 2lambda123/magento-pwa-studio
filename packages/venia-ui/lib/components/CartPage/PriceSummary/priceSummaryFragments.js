import gql from 'graphql-tag';

import { DiscountSummaryFragment } from './discountSummary';
import { GiftCardSummaryFragment } from './giftCardSummary';
import { ShippingSummaryFragment } from './shippingSummary';
import { TaxSummaryFragment } from './taxSummary';

export const PriceSummaryFragment = gql`
    fragment PriceSummaryFragment on Cart {
        id
        items {
            quantity
        }
        ...ShippingSummaryFragment
        prices {
            ...TaxSummaryFragment
            ...DiscountSummaryFragment
            grand_total {
                currency
                value
            }
            subtotal_excluding_tax {
                currency
                value
            }
        }
        ...GiftCardSummaryFragment
    }
    ${DiscountSummaryFragment}
    ${GiftCardSummaryFragment}
    ${ShippingSummaryFragment}
    ${TaxSummaryFragment}
`;
