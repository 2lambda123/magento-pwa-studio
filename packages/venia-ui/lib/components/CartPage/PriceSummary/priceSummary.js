import React from 'react';
import gql from 'graphql-tag';
import { Price } from '@magento/peregrine';
import { usePriceSummary } from '@magento/peregrine/lib/talons/CartPage/PriceSummary/usePriceSummary';
import Button from '../../Button';
import { mergeClasses } from '../../../classify';
import defaultClasses from './priceSummary.css';
import DiscountSummary from './discountSummary';
import GiftCardSummary from './giftCardSummary';
import ShippingSummary from './shippingSummary';
import TaxSummary from './taxSummary';
import { PriceSummaryFragment } from './priceSummaryFragments';

const GET_PRICE_SUMMARY = gql`
    query getPriceSummary($cartId: String!) {
        cart(cart_id: $cartId) {
            id
            ...PriceSummaryFragment
        }
    }
    ${PriceSummaryFragment}
`;

/**
 * A component that fetches and renders cart data including:
 *  - subtotal
 *  - discounts applied
 *  - gift cards applied
 *  - estimated tax
 *  - estimated shipping
 *  - estimated total
 */
const PriceSummary = props => {
    const { isCartUpdating } = props;
    const classes = mergeClasses(defaultClasses, props.classes);
    const talonProps = usePriceSummary({
        queries: {
            getPriceSummary: GET_PRICE_SUMMARY
        }
    });

    const {
        handleProceedToCheckout,
        hasError,
        hasItems,
        isLoading,
        flatData
    } = talonProps;

    if (hasError) {
        return (
            <div className={classes.root}>
                An error occurred. Please refresh the page.
            </div>
        );
    } else if (!hasItems || isLoading) {
        return null;
    }

    const { subtotal, total, discounts, giftCards, taxes, shipping } = flatData;

    const priceClass = isCartUpdating ? classes.priceUpdating : classes.price;
    const totalPriceClass = isCartUpdating
        ? classes.priceUpdating
        : classes.totalPrice;

    return (
        <div className={classes.root}>
            <div className={classes.lineItems}>
                <span className={classes.lineItemLabel}>{'Subtotal'}</span>
                <span className={priceClass}>
                    <Price
                        value={subtotal.value}
                        currencyCode={subtotal.currency}
                    />
                </span>
                <DiscountSummary
                    classes={{
                        lineItemLabel: classes.lineItemLabel,
                        price: priceClass
                    }}
                    data={discounts}
                />
                <GiftCardSummary
                    classes={{
                        lineItemLabel: classes.lineItemLabel,
                        price: priceClass
                    }}
                    data={giftCards}
                />
                <TaxSummary
                    classes={{
                        lineItemLabel: classes.lineItemLabel,
                        price: priceClass
                    }}
                    data={taxes}
                />
                <ShippingSummary
                    classes={{
                        lineItemLabel: classes.lineItemLabel,
                        price: priceClass
                    }}
                    data={shipping}
                />
                <span className={classes.totalLabel}>{'Estimated Total'}</span>
                <span className={totalPriceClass}>
                    <Price value={total.value} currencyCode={total.currency} />
                </span>
            </div>
            <div className={classes.checkoutButton_container}>
                <Button priority={'high'} onClick={handleProceedToCheckout}>
                    {'Proceed to Checkout'}
                </Button>
            </div>
        </div>
    );
};

export default PriceSummary;
