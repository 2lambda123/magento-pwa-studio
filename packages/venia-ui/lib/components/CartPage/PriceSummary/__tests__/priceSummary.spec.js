import React from 'react';
import { createTestInstance } from '@magento/peregrine';

import PriceSummary from '../priceSummary';
import { useLazyQuery } from '@apollo/react-hooks';

jest.mock('../../../../classify');
jest.mock('@apollo/react-hooks', () => {
    const runQuery = jest.fn();
    const queryResult = {
        data: {
            cart: {
                items: [
                    {
                        quantity: 1
                    }
                ],
                applied_gift_cards: [],
                shipping_addresses: [
                    {
                        selected_shipping_method: {
                            amount: {
                                value: 0,
                                currency: 'USD'
                            }
                        }
                    }
                ],
                prices: {
                    subtotal_excluding_tax: {
                        currency: 'USD',
                        value: 11
                    },
                    grand_total: {
                        currency: 'USD',
                        value: 10
                    },
                    discounts: [
                        {
                            amount: {
                                value: 1,
                                currency: 'USD'
                            }
                        }
                    ],
                    applied_taxes: [
                        {
                            amount: {
                                value: 0,
                                currency: 'USD'
                            }
                        }
                    ]
                }
            }
        },
        error: null,
        loading: false
    };
    const useLazyQuery = jest.fn(() => [runQuery, queryResult]);

    return { useLazyQuery };
});

jest.mock('@magento/peregrine/lib/context/cart', () => {
    const state = { cartId: 'cart123' };
    const api = {};
    const useCartContext = jest.fn(() => [state, api]);

    return { useCartContext };
});

jest.mock('@magento/peregrine', () => {
    const Price = props => <span>{`$${props.value}`}</span>;

    return {
        ...jest.requireActual('@magento/peregrine'),
        Price
    };
});

test('renders PriceSummary correctly', () => {
    const tree = createTestInstance(<PriceSummary />);

    expect(tree.toJSON()).toMatchSnapshot();
});

test('renders an error state if query fails', () => {
    useLazyQuery.mockReturnValueOnce([
        jest.fn(),
        {
            error: true
        }
    ]);

    const tree = createTestInstance(<PriceSummary />);

    expect(tree.toJSON()).toMatchSnapshot();
});

test('renders nothing if query is loading', () => {
    useLazyQuery.mockReturnValueOnce([
        jest.fn(),
        {
            loading: true
        }
    ]);

    const tree = createTestInstance(<PriceSummary />);

    expect(tree.toJSON()).toMatchSnapshot();
});
test('renders nothing if query returns no items', () => {
    useLazyQuery.mockReturnValueOnce([
        jest.fn(),
        {
            data: {
                cart: {
                    items: [
                        // Intentionally empty
                    ],
                    applied_gift_cards: [],
                    shipping_addresses: [
                        {
                            selected_shipping_method: {
                                amount: {
                                    value: 0,
                                    currency: 'USD'
                                }
                            }
                        }
                    ],
                    prices: {
                        subtotal_excluding_tax: {
                            currency: 'USD',
                            value: 11
                        },
                        grand_total: {
                            currency: 'USD',
                            value: 10
                        },
                        discounts: [
                            {
                                amount: {
                                    value: 1,
                                    currency: 'USD'
                                }
                            }
                        ],
                        applied_taxes: [
                            {
                                amount: {
                                    value: 0,
                                    currency: 'USD'
                                }
                            }
                        ]
                    }
                }
            },
            error: null,
            loading: false
        }
    ]);

    const tree = createTestInstance(<PriceSummary />);

    expect(tree.toJSON()).toMatchSnapshot();
});
