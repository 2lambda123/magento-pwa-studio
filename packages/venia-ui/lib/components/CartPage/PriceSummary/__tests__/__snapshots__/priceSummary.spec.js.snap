// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders PriceSummary correctly on cart page 1`] = `
<div
  className="root"
>
  <div
    className="lineItems"
  >
    <span
      className="lineItemLabel"
    >
      Subtotal
    </span>
    <span
      className="price"
    >
      <span>
        $3.5
      </span>
    </span>
    <span
      className="totalLabel"
    >
      Estimated Total
    </span>
    <span
      className="totalPrice"
    >
      <span>
        $8.5
      </span>
    </span>
  </div>
  <div
    className="checkoutButton_container"
  >
    <button
      className="root_highPriority"
      disabled={false}
      type="button"
    >
      <span
        className="content"
      >
        Proceed to Checkout
      </span>
    </button>
  </div>
</div>
`;

exports[`renders PriceSummary correctly on checkout page 1`] = `
<div
  className="root"
>
  <div
    className="lineItems"
  >
    <span
      className="lineItemLabel"
    >
      Subtotal
    </span>
    <span
      className="price"
    >
      <span>
        $3.5
      </span>
    </span>
    <span
      className="totalLabel"
    >
      Total
    </span>
    <span
      className="totalPrice"
    >
      <span>
        $8.5
      </span>
    </span>
  </div>
</div>
`;

exports[`renders an error state if query fails 1`] = `
<div
  className="root"
>
  An error occurred. Please refresh the page.
</div>
`;

exports[`renders nothing if query is loading 1`] = `null`;

exports[`renders nothing if query returns no items 1`] = `null`;
